File: lernfeld_01_grundlagen/python/loesungen/aufgabe_03.py
"""Aufgabe 3: Zahlenraten — Musterlösung (Python).

Zufallszahl zwischen 1 und 100 erraten, mit Hinweisen, Versuchszähler,
„Noch eine Runde?"-Abfrage und Runden-Statistik (Bonus).
"""

import random


def spiele_runde() -> int:
    """Spielt eine Runde Zahlenraten und gibt die Anzahl der Versuche zurück."""
    geheim = random.randint(1, 100)
    versuche = 0

    print("Ich habe eine Zahl zwischen 1 und 100 gewählt.")
    while True:
        eingabe = input("Dein Tipp: ").strip()
        try:
            tipp = int(eingabe)
        except ValueError:
            print("Bitte eine ganze Zahl eingeben.")
            continue  # ungültige Eingaben zählen nicht als Versuch

        versuche += 1
        if tipp < geheim:
            print("Zu klein!")
        elif tipp > geheim:
            print("Zu groß!")
        else:
            print(f"Richtig! Die Zahl war {geheim}.")
            print(f"Du hast {versuche} Versuche gebraucht.")
            return versuche


def main() -> None:
    runden = []

    while True:
        runden.append(spiele_runde())

        nochmal = input("Noch eine Runde? (j/n): ").strip().lower()
        if nochmal != "j":
            break

    # Statistik aller Runden (Bonus)
    print(f"\nDu hast {len(runden)} Runde(n) gespielt.")
    print(f"Beste Runde: {min(runden)} Versuch(e)")
    print(f"Durchschnitt: {sum(runden) / len(runden):.1f} Versuche")
    print("Danke fürs Spielen!")


if __name__ == "__main__":
    main()

File: lernfeld_01_grundlagen/cpp/loesungen/aufgabe_03.cpp
// Aufgabe 3: Zahlenraten — Musterlösung (C++)
//
// Zufallszahl zwischen 1 und 100 erraten, mit Hinweisen, Versuchszähler
// und „Noch eine Runde?"-Abfrage. Nutzt modernes <random> statt rand().
//
// Kompilieren:  g++ -std=c++17 -Wall -Wextra aufgabe_03.cpp -o aufgabe_03
// Ausführen:    ./aufgabe_03

#include <iostream>  // std::cout, std::cin
#include <limits>    // std::numeric_limits
#include <random>    // std::random_device, std::mt19937, uniform_int_distribution
#include <string>    // std::string

// Spielt eine Runde und gibt die Anzahl der Versuche zurück.
int spiele_runde() {
    // Zufallsgenerator (echter Zufallsstart + Mersenne-Twister)
    std::random_device rd;
    std::mt19937 generator(rd());
    std::uniform_int_distribution<int> verteilung(1, 100);
    const int geheim = verteilung(generator);

    int versuche = 0;
    int tipp;
    std::cout << "Ich habe eine Zahl zwischen 1 und 100 gewählt." << std::endl;

    while (true) {
        std::cout << "Dein Tipp: ";
        std::cin >> tipp;

        if (std::cin.fail()) {
            std::cin.clear();
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
            std::cout << "Bitte eine ganze Zahl eingeben." << std::endl;
            continue;  // ungültige Eingaben zählen nicht als Versuch
        }

        ++versuche;
        if (tipp < geheim) {
            std::cout << "Zu klein!" << std::endl;
        } else if (tipp > geheim) {
            std::cout << "Zu groß!" << std::endl;
        } else {
            std::cout << "Richtig! Die Zahl war " << geheim << "." << std::endl;
            std::cout << "Du hast " << versuche << " Versuche gebraucht."
                      << std::endl;
            return versuche;
        }
    }
}

int main() {
    int runden = 0;
    int versuche_gesamt = 0;
    int beste_runde = 0;

    while (true) {
        const int versuche = spiele_runde();
        ++runden;
        versuche_gesamt += versuche;
        if (beste_runde == 0 || versuche < beste_runde) {
            beste_runde = versuche;
        }

        std::string nochmal;
        std::cout << "Noch eine Runde? (j/n): ";
        std::cin >> nochmal;
        if (nochmal != "j" && nochmal != "J") {
            break;
        }
    }

    // Statistik aller Runden (Bonus)
    std::cout << "\nDu hast " << runden << " Runde(n) gespielt." << std::endl;
    std::cout << "Beste Runde: " << beste_runde << " Versuch(e)" << std::endl;
    std::cout << "Durchschnitt: " << static_cast<double>(versuche_gesamt) / runden
              << " Versuche" << std::endl;
    std::cout << "Danke fürs Spielen!" << std::endl;

    return 0;
}
